### DISCUSSION TITLE
How would one run osxphotos from Python interactive shell?

### DISCUSSION THREAD

**Discussion by hohealpen** [RhetTbull/osxphotos#822] [Q&A]
Killer library - the `osxphotos inspect` is nifty - by enabling correlating the photo's GUID to original filename, it's the first true workaround to the ghosted menu "Show in Finder"

### Question
Do you have a way you would `import osxphotos` in a python shell and interact with the API?

Reason: before scripting, I usually have a play with the API in [bpython](https://bpython-interpreter.org/) because it has a killer API autocompleter. I've had some version compatibility issues with bpython/homebrew/python/pipx, but that aside, I've had trouble even doing `import osxphotos` from a plain python shell. I'm not sure what to put in PYTHONPATH or sys.path to get the interpreter to find the module. Only when the import is in a script called from the CLI `osxphotos run <script.py>` does it work.

I'm sure I'm being daft and missing something obvious, let me know how you might do it.

**Comment by RhetTbull** [comment]
Hi. Not sure what the issue is with importing osxphotos. As long as osxphotos is installed in your environment you should be able to directly import it. However, for just playing around or for prototyping, osxphotos includes a built in Python REPL based on [ptpython](https://github.com/prompt-toolkit/ptpython) (similar to bpython) that has great auto-complete. Just type `osxphotos repl`.  The REPL also includes an interactive object explorer (e.g. type `explore(PhotoInfo)` to explore all properties and methods of the PhotoInfo class.  

You can use this REPL even if python isn't installed system-wide as it uses the python environment installed with osxphotos. When you launch `osxphotos repl` also preloads several variables (e..g the `photos` variable will contain a list of all your photos as `PhotoInfo` objects).  

If you find you need an external package from PyPI in your script, you can also use `osxphotos install package_name` which does a `pip install` of the package into the osxphotos environment.  

In short, osxphotos is not just a python app but also a full python environment that gives you full access to python while using osxphotos.

**Reply by RhetTbull** [reply]
For reference, here's an example output of `osxphotos repl` with a test library:

```pycon
$ osxphotos repl
python version: 3.9.13 (main, Nov  6 2022, 18:33:56)
[Clang 12.0.0 (clang-1200.0.32.29)]
osxphotos version: 0.52.0
Using last opened Photos library: /Users/rhet/Pictures/Test-10.15.7.photoslibrary
Loading database
Processing database /Users/rhet/Pictures/Test-10.15.7.photoslibrary/database/photos.db
Processing database /Users/rhet/Pictures/Test-10.15.7.photoslibrary/database/Photos.sqlite
Processing database.
Database version: 6000, 5.
Processing persons in photos.
Processing detected faces in photos.
Processing albums.
Processing keywords.
Processing photo details.
Processing import sessions.
Processing additional photo details.
Processing face details.
Processing photo labels.
Processing EXIF details.
Processing computed aesthetic scores.
Processing comments and likes for shared photos.
Processing moments.
Done processing details from Photos library.
Done: took 0.14 seconds
Getting photos
Found 63 photos in 0.00 seconds

The following classes have been imported from osxphotos:
- AlbumInfo, ExifTool, PhotoInfo, PhotoExporter, ExportOptions, ExportResults, PhotosDB, PlaceInfo, QueryOptions, MomentInfo, ScoreInfo, SearchInfo

The following variables are defined:
- photosdb: PhotosDB() instance for '/Users/rhet/Pictures/Test-10.15.7.photoslibrary'
- photos: list of PhotoInfo objects for all photos filtered with any query options passed on command line (len=63)
- all_photos: list of PhotoInfo objects for all photos in photosdb, including those in the trash (len=65)
- selected: list of PhotoInfo objects for any photos selected in Photos (len=0)

The following functions may be helpful:
- get_photo(uuid): return a PhotoInfo object for photo with uuid; e.g. get_photo('B13F4485-94E0-41CD-AF71-913095D62E31')
- get_selected(); return list of PhotoInfo objects for photos selected in Photos
- show(photo): open a photo object in the default viewer; e.g. show(selected[0])
- show(path): open a file at path in the default viewer; e.g. show('/path/to/photo.jpg')
- spotlight(photo): open a photo and spotlight it in Photos
- inspect(object): print information about an object; e.g. inspect(PhotoInfo)
- explore(object): interactively explore an object with objexplore; e.g. explore(PhotoInfo)
- q, quit, quit(), exit, exit(): exit this interactive shell

>>>
```

**Comment by RhetTbull** [comment]
It occurred to me you may have installed osxphotos using `pipx` then tried to do `import osxphotos` from your system python prompt. If so, that won't work as `pipx` creates a separate python environment for each app. You'd need to do `python3 -m pip install osxphotos` to install osxphotos into your python environment then you could do `import osxphotos`.  But, as stated above, it may be sufficient to use the built in REPL for osxphotos using `osxphotos repl`.  I use this all the time for experimenting and prototyping.

**Comment by hohealpen** [comment]
The built-in REPL looks completely suitable for purpose - great.

Thanks for the example, worked for me as advertised (a little slow to
initialize against a 60,000 image library, so will look to see if it can
lazy load anything)

I did use pipx to install it, so yes, your theory sounds right to me. I
was/am tripping over the issue too where homebrew installs python3.11 when
it installs bpython or pipx, but osxphotos uses 3.10, but again, while I
like bpython, the built-in REPL is just fine - it's only to have a look at
object's methods and have a play.

I plan to use your library to allow me to say, look at a photo's "native"
name in Photos, then have your software tell me to which GUID-based name it
relates when I'm browsing the images in Luminar Neo... Ukrainian Lightroom
:-) I have Luminar pointed at the Photos Library.photoslibrary/originals
(Skylum devs don't advise it, but copying massive sets of images to another
directory defeats the advantage of a lightweight, non-destructive editor,
no?). Because Luminar is looking at ./originals, it sees only the GUID
filenames because it doesn't have your library to correlate them to the the
Photos metadata.

Thanks again for some very clever software and prompt, useful replies, Rhet.

Cheers -

Geo

**Reply by RhetTbull** [reply]
> so will look to see if it can lazy load anything

Unfortunately it cannot.  osxphotos version 0.54.0, released yesterday, does work with Python 3.11 so you can install that in your system python if desired.

> I plan to use your library to allow me to say, look at a photo's "native"
> name in Photos, then have your software tell me to which GUID-based name it
> relates when I'm browsing the images in Luminar Neo.

If that's your primary purpose, you can easily export a copy of your library with osxphotos using hardlinks which will map the exported files directly to the originals in the library but with readable names.  For example the following command:

`osxphotos export /path/to/export--export-as-hardlink --directory "{created.year}/{created.mm}/{created.dd}" --filename "{original_name}" --skip-edited`

This doesn't take up any space on the disk because it uses hardlinks (but the export path must be on the same volume as the Photos library), uses a directory naming scheme in the form of `2022/11/13` based on photo creation date and uses the original file name of the photo (e.g. `IMG1234.jpeg`) instead of the GUID.  The `--skip-edited` flag exports only originals, not edited images.
